### PR TITLE
Added module support, and available Astilectron and Electron

### PR DESCRIPTION
**Pull request by alexj212**
This will allow go-astilectron-bundler to config the version of Electron and Astilectron that are to be downloaded and used for runtime. There will be a corresponding PR in go-astilectron-bundler.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,3 +3,4 @@ Thumbs.db
 .idea/
 cover*
 testdata/tmp/*
+go.sum
```
Comment (asticode) [inline .gitignore]: I'd rather not complexify the PR with adding module support. Can you remove this and make it in another PR?

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,17 @@
+
```
Comment (asticode) [inline Makefile]: I'd rather not complexify the PR with updating the Makefile. Can you remove this and make it in another PR?

File: README.md [hunk 3]
```diff
@@ -56,6 +56,8 @@ var a, _ = astilectron.New(astilectron.Options{
     AppIconDefaultPath: "<your .png icon>", // If path is relative, it must be relative to the data directory
     AppIconDarwinPath:  "<your .icns icon>", // Same here
     BaseDirectoryPath: "<where you want the provisioner to install the dependencies>",
+	VersionAstilectron: "0.33.0",
```
Comment (asticode) [inline README.md]: Can you align with the line above?

File: astilectron.go [hunk 4]
```diff
@@ -20,8 +20,8 @@ import (
 // Versions
 const (
 	DefaultAcceptTCPTimeout = 30 * time.Second
-	VersionAstilectron      = "0.33.0"
-	VersionElectron         = "4.0.1"
+	DefaultVersionAstilectron      = "0.33.0"
```
Comment (asticode) [inline astilectron.go]: Can you run `gofmt` ?

File: go.mod [hunk 5]
```diff
@@ -0,0 +1,11 @@
+module github.com/asticode/go-astilectron
```
Comment (asticode) [inline go.mod]: I'd rather not complexify the PR with adding module support. Can you remove this and make it in another PR?

File: paths.go [hunk 6]
```diff
@@ -27,12 +27,24 @@ type Paths struct {
 	electronUnzipSrc       string
 	provisionStatus        string
 	vendorDirectory        string
+	VersionAstilectron     string
```
Comment (asticode) [inline paths.go]: Versions shouldn't be stored in `Paths`. Can you remove these attributes?

File: astilectron.go [hunk 7]
```diff
@@ -93,6 +95,15 @@ func New(o Options) (a *Astilectron, err error) {
 		return
 	}
 
+	if o.VersionAstilectron == "" {
+		o.VersionAstilectron = DefaultVersionAstilectron
+	}
+
```
Comment (asticode) [inline astilectron.go]: Can you remove this empty line so that we group the 2 `if` ?

File: paths.go [hunk 8]
```diff
@@ -27,12 +27,24 @@ type Paths struct {
...
+	VersionElectron        string
 }
 
 // newPaths creates new paths
 func newPaths(os, arch string, o Options) (p *Paths, err error) {
+	if o.VersionAstilectron == "" {
```
Comment (asticode) [inline paths.go]: Unless I missed something, this should be removed. Indeed, defaults are already set in `Astilectron`'s `New()` method.

File: paths.go [hunk 9]
```diff
@@ -27,12 +27,24 @@ type Paths struct {
...
+		o.VersionAstilectron = DefaultVersionAstilectron
+	}
+	if o.VersionElectron == "" {
+		o.VersionElectron = DefaultVersionElectron
+	}
+
+
 	// Init base directory path
 	p = &Paths{}
+	p.VersionAstilectron = o.VersionAstilectron
```
Comment (asticode) [inline paths.go]: You need to remove theses lines as version shouldn't be stored in `Paths`

File: paths.go [hunk 10]
```diff
@@ -114,12 +126,12 @@ func (p *Paths) initDataDirectory(dataDirectoryPath, appName string) (err error)
 }
 
 // AstilectronDownloadSrc returns the download URL of the (currently platform-independent) astilectron zip file
-func AstilectronDownloadSrc() string {
+func AstilectronDownloadSrc( VersionAstilectron string) string {
```
Comment (asticode) [inline paths.go]: Variable name should begin with a lower case letter : `versionAstilectron`

File: paths.go [hunk 11]
```diff
@@ -114,12 +126,12 @@ func (p *Paths) initDataDirectory(dataDirectoryPath, appName string) (err error)
...
 	return fmt.Sprintf("https://github.com/asticode/astilectron/archive/v%s.zip", VersionAstilectron)
 }
 
 // ElectronDownloadSrc returns the download URL of the platform-dependant electron zipfile
-func ElectronDownloadSrc(os, arch string) string {
+func ElectronDownloadSrc(os, arch, VersionElectron string) string {
```
Comment (asticode) [inline paths.go]: Variable name should begin with a lower case letter : `versionElectron`

File: paths_test.go [hunk 12]
```diff
@@ -12,12 +12,17 @@ import (
 
 func TestPaths(t *testing.T) {
 	const k = "APPDATA"
+	const VersionAstilectron = DefaultVersionAstilectron
```
Comment (asticode) [inline paths_test.go]: These lines are useless if I'm not mistaken

File: paths_test.go [hunk 13]
```diff
@@ -12,12 +12,17 @@ import (
...
+	const VersionElectron = DefaultVersionElectron
+
 	ad := os.Getenv(k)
 	os.Setenv(k, "")
 	ep, err := os.Executable()
 	ep = filepath.Dir(ep)
 	assert.NoError(t, err)
-	p, err := newPaths("linux", "amd64", Options{})
+
+	o := Options{VersionAstilectron: DefaultVersionAstilectron, VersionElectron: DefaultVersionElectron}
```
Comment (asticode) [inline paths_test.go]: Can you not create this `o` variable and put the `Options` struct directly in the `newPath` call below?

File: provisioner.go [hunk 14]
```diff
@@ -73,14 +73,14 @@ func (p *defaultProvisioner) Provision(ctx context.Context, appName, os, arch st
 		err = errors.Wrap(err, "provisioning astilectron failed")
 		return
 	}
-	s.Astilectron = &ProvisionStatusPackage{Version: VersionAstilectron}
+	s.Astilectron = &ProvisionStatusPackage{Version: paths.VersionAstilectron}
```
Comment (asticode) [inline provisioner.go]: I don't really like the idea of storing the versions in the `Paths` struct.

I think the better solution would be to update the `Provisioner` interface into `Provision(ctx context.Context, appName, os, arch, versionAstilectron, versionElectron string, p Paths) error`.

You would need to update more stuff.

File: provisioner.go [hunk 15]
```diff
@@ -73,14 +73,14 @@ func (p *defaultProvisioner) Provision(ctx context.Context, appName, os, arch st
...
 
 	// Provision electron
 	if err = p.provisionElectron(ctx, paths, s, appName, os, arch); err != nil {
 		err = errors.Wrap(err, "provisioning electron failed")
 		return
 	}
-	s.Electron[provisionStatusElectronKey(os, arch)] = &ProvisionStatusPackage{Version: VersionElectron}
+	s.Electron[provisionStatusElectronKey(os, arch)] = &ProvisionStatusPackage{Version: paths.VersionElectron}
```
Comment (asticode) [inline provisioner.go]: Same as above

File: provisioner.go [hunk 16]
```diff
@@ -144,12 +144,12 @@ func (p *defaultProvisioner) updateProvisionStatus(paths Paths, s *ProvisionStat
 
 // provisionAstilectron provisions astilectron
 func (p *defaultProvisioner) provisionAstilectron(ctx context.Context, paths Paths, s ProvisionStatus) error {
-	return p.provisionPackage(ctx, paths, s.Astilectron, p.moverAstilectron, "Astilectron", VersionAstilectron, paths.AstilectronUnzipSrc(), paths.AstilectronDirectory(), nil)
+	return p.provisionPackage(ctx, paths, s.Astilectron, p.moverAstilectron, "Astilectron", paths.VersionAstilectron, paths.AstilectronUnzipSrc(), paths.AstilectronDirectory(), nil)
```
Comment (asticode) [inline provisioner.go]: Same as above

File: provisioner.go [hunk 17]
```diff
@@ -144,12 +144,12 @@ func (p *defaultProvisioner) updateProvisionStatus(paths Paths, s *ProvisionStat
...
 }
 
 // provisionElectron provisions electron
 func (p *defaultProvisioner) provisionElectron(ctx context.Context, paths Paths, s ProvisionStatus, appName, os, arch string) error {
-	return p.provisionPackage(ctx, paths, s.Electron[provisionStatusElectronKey(os, arch)], p.moverElectron, "Electron", VersionElectron, paths.ElectronUnzipSrc(), paths.ElectronDirectory(), func() (err error) {
+	return p.provisionPackage(ctx, paths, s.Electron[provisionStatusElectronKey(os, arch)], p.moverElectron, "Electron", paths.VersionElectron, paths.ElectronUnzipSrc(), paths.ElectronDirectory(), func() (err error) {
```
Comment (asticode) [inline provisioner.go]: Same as above

File: provisioner_test.go [hunk 18]
```diff
@@ -18,7 +18,7 @@ func testProvisionerSuccessful(t *testing.T, p Paths, osName, arch string) {
 	assert.NoError(t, err)
 	b, err := ioutil.ReadFile(p.ProvisionStatus())
 	assert.NoError(t, err)
-	assert.Equal(t, "{\"astilectron\":{\"version\":\""+VersionAstilectron+"\"},\"electron\":{\""+provisionStatusElectronKey(osName, arch)+"\":{\"version\":\""+VersionElectron+"\"}}}\n", string(b))
+	assert.Equal(t, "{\"astilectron\":{\"version\":\""+p.VersionAstilectron+"\"},\"electron\":{\""+provisionStatusElectronKey(osName, arch)+"\":{\"version\":\""+p.VersionElectron+"\"}}}\n", string(b))
```
Comment (asticode) [inline provisioner_test.go]: Versions will need to be passed in the `testProvisionerSuccessful` parameters instead

File: .gitignore (hunk 1 above)
Comment (alexj212) [reply .gitignore]: Removed additions of go.mod, .gitignore and Makefile - will leave for future PR.

File: Makefile (hunk 2 above)
Comment (alexj212) [reply Makefile]: Removed additions of go.mod, .gitignore and Makefile - will leave for future PR.

File: README.md (hunk 3 above)
Comment (alexj212) [reply README.md]: Fixed

File: astilectron.go (hunk 4 above)
Comment (alexj212) [reply astilectron.go]: gofmt run on src

File: go.mod (hunk 5 above)
Comment (alexj212) [reply go.mod]: Removed additions of go.mod, .gitignore and Makefile - will leave for future PR.

File: provisioner.go (hunk 14 above)
Comment (alexj212) [reply provisioner.go]: Removed storing of versions in Paths. Updated function signatures, to pass versions

File: provisioner.go (hunk 15 above)
Comment (alexj212) [reply provisioner.go]: Update as above comment

File: provisioner.go (hunk 16 above)
Comment (alexj212) [reply provisioner.go]: Update as above comment

File: provisioner.go (hunk 17 above)
Comment (alexj212) [reply provisioner.go]: Update as above comment

File: provisioner_test.go (hunk 18 above)
Comment (alexj212) [reply provisioner_test.go]: Update as above comment

**Comment by alexj212** [comment]
Updated with changes requested.

File: paths.go (hunk 8 above)
Comment (alexj212) [reply paths.go]: removed - I had added due to passing empty Options{} from tests. Making sure to populate Options{} in the tests with default versions.

File: paths.go (hunk 10 above)
Comment (alexj212) [reply paths.go]: Fixed

File: paths.go (hunk 11 above)
Comment (alexj212) [reply paths.go]: fixed

File: paths_test.go (hunk 12 above)
Comment (alexj212) [reply paths_test.go]: Left as test reference VersionAstilectron from when it was a global.

File: paths_test.go (hunk 13 above)
Comment (alexj212) [reply paths_test.go]: Left since Options{} being passed to newPaths() would cause issue since versions would not be set. reused Options o passing to multiple tests.

File: paths_test.go (hunk 12 above)
Comment (asticode) [reply paths_test.go]: Since the behavior has changed I'd rather remove these lines and use `DefaultVersionAstilectron` and `DefaultVersionElectron` below

**Comment by alexj212** [comment]
Added the changes for DefaultVersionElectron and DefaultVersionAstilectron

**Comment by asticode** [comment]
I think you forgot to push your last commit but that's no problem, I'll merge the PR and update the code afterwards.

Thanks!
